src/wallet/rpc/spend.cpp

```diff
@@ -1060,101 +1159,248 @@ RPCHelpMan send()
+            // estimate final size of tx
+            TxSize tx_size = CalculateMaximumSignedTxSize(CTransaction(rawTx), pwallet.get());
+            CAmount fee_from_size = fee_rate.GetFee(tx_size.vsize);
+            CAmount effective_value = total_input_value - fee_from_size;
+
+            if (effective_value <= 0) {
+                if (sendmax) {
+                    throw JSONRPCError(RPC_WALLET_INSUFFICIENT_FUNDS, "Total value of UTXO pool too low to pay for sweep, try using lower feerate.");
                 } else {
-                    result.pushKV("hex", hex);
+                    throw JSONRPCError(RPC_WALLET_INSUFFICIENT_FUNDS, "Total value of UTXO pool too low to pay for sweep. Try using lower feerate or excluding uneconomic UTXOs with 'sendmax' option.");
+                }
+            }
+
+            CAmount output_amounts_claimed(0);
+            for (CTxOut out : rawTx.vout) {
+                output_amounts_claimed += out.nValue;
+            }
+
+            if (output_amounts_claimed > total_input_value) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Assigned more value to outputs than available funds.");
+            }
+
+            CAmount remainder = effective_value - output_amounts_claimed;
+            if (remainder < 0) {
+                throw JSONRPCError(RPC_WALLET_INSUFFICIENT_FUNDS, "Insufficient funds for fees after creating specified outputs.");
+            }
+
+            CAmount per_output_without_amount = remainder / addresses_without_amount.size();
+
+            bool gave_remaining_to_first = false;
+            for (CTxOut& out : rawTx.vout) {
+                CTxDestination dest;
+                ExtractDestination(out.scriptPubKey, dest);
+                std::string addr = EncodeDestination(dest);
+                if (addresses_without_amount.count(addr) > 0) {
+                    out.nValue = per_output_without_amount;
+                    if (!gave_remaining_to_first) {
+                        out.nValue += remainder % addresses_without_amount.size();
+                        gave_remaining_to_first = true;
+                    }
+                    if (IsDust(out, pwallet->chain().relayDustFee())) {
+                        // Dynamically generated output amount is dust
+                        throw JSONRPCError(RPC_WALLET_INSUFFICIENT_FUNDS, "Dynamically assigned remainder results in dust output.");
+                    }
+                } else {
+                    if (IsDust(out, pwallet->chain().relayDustFee())) {
+                        // Specified output amount is dust
+                        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Specified output amount to %s is below dust threshold.", addr));
+                    }
+                }
+            }
+
+            if (lock_unspents) {
+                for (const CTxIn& txin : rawTx.vin) {
+                    pwallet->LockCoin(txin.prevout);
```

Comment:
b2b1d6b minor suggestions if you retouch and are so inclined

<details><summary>diff</summary><p>

```diff
@@ -1353,17 +1353,11 @@ RPCHelpMan sweepwallet()
             }
 
             CCoinControl coin_control;
-
-            SetFeeEstimateMode(*pwallet, coin_control, options["conf_target"], options["estimate_mode"], options["fee_rate"], /* override_min_fee */ false);
+            SetFeeEstimateMode(*pwallet, coin_control, options["conf_target"], options["estimate_mode"], options["fee_rate"], /*override_min_fee=*/false);
 
             coin_control.fAllowWatchOnly = ParseIncludeWatchonly(options["include_watching"], *pwallet);
 
-            bool lock_unspents = false;
-            if (options.exists("lock_unspents")) {
-                lock_unspents = options["lock_unspents"].get_bool();
-            }
-
-            bool rbf = options.exists("replaceable") ? options["replaceable"].get_bool() : pwallet->m_signal_rbf;
+            const bool rbf{options.exists("replaceable") ? options["replaceable"].get_bool() : pwallet->m_signal_rbf};
 
             FeeCalculation fee_calc_out;
             CFeeRate fee_rate = GetMinimumFeeRate(*pwallet, coin_control, &fee_calc_out);
@@ -1422,7 +1416,7 @@ RPCHelpMan sweepwallet()
                 }
             }
 
-            CAmount output_amounts_claimed(0);
+            CAmount output_amounts_claimed{0};
             for (CTxOut out : rawTx.vout) {
                 output_amounts_claimed += out.nValue;
             }
@@ -1431,12 +1425,12 @@ RPCHelpMan sweepwallet()
                 throw JSONRPCError(RPC_INVALID_PARAMETER, "Assigned more value to outputs than available funds.");
             }
 
-            CAmount remainder = effective_value - output_amounts_claimed;
+            const CAmount remainder{effective_value - output_amounts_claimed};
             if (remainder < 0) {
                 throw JSONRPCError(RPC_WALLET_INSUFFICIENT_FUNDS, "Insufficient funds for fees after creating specified outputs.");
             }
 
-            CAmount per_output_without_amount = remainder / addresses_without_amount.size();
+            const CAmount per_output_without_amount = remainder / addresses_without_amount.size();
 
             bool gave_remaining_to_first = false;
             for (CTxOut& out : rawTx.vout) {
@@ -1461,6 +1455,7 @@ RPCHelpMan sweepwallet()
                 }
             }
 
+            const bool lock_unspents{options.exists("lock_unspents") ? options["lock_unspents"].get_bool() : false};
             if (lock_unspents) {
                 for (const CTxIn& txin : rawTx.vin) {
                     pwallet->LockCoin(txin.prevout);
```
</p></details>

Reply:
I think I got all these.